### ISSUE TITLE
_ for Unused Parameters

### ISSUE THREAD

**Issue by User0-0** [dart-lang/language#204]
What about implementing this feature dart-lang/sdk/issues/4019?

**Comment by User0-0** [comment]
@munificent I'm sure you had seen the full comment including,
> Maybe it can made as `( , , , ...)`(without any name) or with any other symbol so it doesn't effect the language styling or whatever.

What about using a symbol which which can't be used as identifiers or without any identifiers `( , , ,...)`?

**Comment by andreashaese** [comment]
> However, it is technically a breaking change.

Does that mean that the proposal is rejected in its current form?

> There may be existing code that uses `_` as a parameter name and actually accesses it later. We are obviously very careful about breaking users' existing code.

The motivation is clear, but that seems a little vague. Do you have evidence that this would cause unacceptably much actual harm? E.g. by counting occurrences of such patterns in code repositories?

There is precedent for using `_` as generic "don't care" placeholder in other languages (Rust, Swift, ...), which is why I like the original proposal. But right now I don't know whether minds are already made up.

**Comment by munificent** [comment]
> Does that mean that the proposal is rejected in its current form?

No, things are rarely that black and white. We aren't looking at this proposal right now because we're focused on other things (mainly non-nullable types).

> Do you have evidence that this would cause unacceptably much actual harm? E.g. by counting occurrences of such patterns in code repositories?

I don't have that data now, but that is how we generally investigate language changes like this, yes. Of course, gathering this data takes time too. :)

> There is precedent for using `_` as generic "don't care" placeholder in other languages (Rust, Swift, ...), which is why I like the original proposal.

Yup, that's one of the reasons we like it too. If we add pattern matching and destructuring to Dart, then we will very likely want `_` patterns. At that point, it begins to make more sense to treat `_` as non-binding in other contexts. So it may be that we revisit this proposal when/if we look at pattern matching.

**Comment by andreashaese** [comment]
Thanks for the clarification! Sorry for being off-topic: Is there a roadmap where I can review when such a proposal would fit into the schedule, or even help making it a priority? That would help me focus on the relevant discussions.

**Comment by munificent** [comment]
> Sorry for being off-topic: Is there a roadmap where I can review when such a proposal would fit into the schedule, or even help making it a priority?

We don't have a long-term roadmap written down anywhere right now. What we know so far is:

* In the short-term, we're implementing the new changes to collections: set literals, spread, `if`, and `for`.

* Concurrently and and after that, most of the team is focused on non-nullable types. Because it's helpful for evolving the core libraries to work better with non-nullable types, this might involve extension methods too.

Beyond that, nothing is pinned down.

**Comment by mit-mit** [comment]
The best current approximation for a roadmap is https://github.com/dart-lang/language/projects/1

**Comment by andreashaese** [comment]
Thank you both very much!

Back on topic, I'll add a reference to #207 since as @munificent mentioned, both proposals intersect.

**Comment by natebosch** [comment]
> However, it is technically a breaking change. There may be existing code that uses `_` as a parameter name and actually accesses it later.

Would it be sensible to say that `_` can be accessed if it has only 1 definition, and is an error otherwise? Existing code must only define it once and so it's behavior wouldn't change.

**Comment by andreashaese** [comment]
Purely my opinion: reading from `_` to me is an anti pattern even in currently existing code, and I would be surprised if it was broadly used like this today. After all, variables have names _because_ that expresses their meaning within the context in which you're going to access them. `_` may not formally have that "don't care" semantic so far, but it's already now treated specially by the compiler in that it doesn't generate a warning if it's only ever written to, but never read from.

Not sure if that's viable, but a migrator could replace `_` instances that are read from with `_1`, `_2` etc., or interactively ask which new name to use.

**Comment by davidmorgan** [comment]
We could flag accessing `_` with a hint/lint if there's just one, and an error if there's more than one.

**Comment by vanesyan** [comment]
> We could flag accessing `_` with a hint/lint if there's just one, and an error if there's more than one.

Why should it erroring on multiple use? It will be very helpful in some cases in destructuring arrays, etc. (if it's landed) to ignore more than one values, e.g.:
```dart
List<dynamic> foo() {
  return [data, someData, wantedData];
}

final [_, _, a] = foo();
```

**Comment by andreashaese** [comment]
Yes, this is a key aspect of the proposal. What @natebosch and @davidmorgan suggest (please correct me if I misunderstood!) is that _reading_ from `_` would emit a warning. This is to ensure that existing code doesn't break, but when compiled, developers learn that it's discouraged. Reading `_` when there are ambiguous declarations (which would then be allowed, like in your example) would of course be a hard error as expected.

**Comment by munificent** [comment]
> Why should it erroring on multiple use?

By "use", @davidmorgan means *reading* the `_` variable, not declaring it.
